### PR TITLE
Fix wrong hint when status checking is running on pull request view

### LINKED ISSUE #9394: Status check on pull view is not correct.
图片

### PR DESCRIPTION
**Pull request by lunny**
When all required status check were running, it will display some required status check failed.

This PR will fix that. There are more states to check about the status check.

Fix #9394

### CODE REVIEW

File: routers/repo/pull.go [hunk 1]
```diff
@@ -417,7 +418,9 @@ func PrepareViewPullInfo(ctx *context.Context, issue *models.Issue) *git.Compare
 			}
 			return false
 		}
-		ctx.Data["IsRequiredStatusCheckSuccess"] = pull_service.IsCommitStatusContextSuccess(commitStatuses, pull.ProtectedBranch.StatusCheckContexts)
+		state := pull_service.MergeRequiredContextsCommitStatus(commitStatuses, pull.ProtectedBranch.StatusCheckContexts)
+		ctx.Data["RequiredStatusCheckState"] = state
+		ctx.Data["IsRequiredStatusCheckSuccess"] = state == structs.CommitStatusSuccess
```
Comment (guillep2k) [inline routers/repo/pull.go]:
```suggestion
		ctx.Data["IsRequiredStatusCheckSuccess"] = state.IsSuccess()
```

File: modules/structs/commit_status.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	// CommitStatusError is for when the Status is Error
+	CommitStatusError CommitStatusState = "error"
+	// CommitStatusFailure is for when the Status is Failure
+	CommitStatusFailure CommitStatusState = "failure"
+	// CommitStatusWarning is for when the Status is Warning
+	CommitStatusWarning CommitStatusState = "warning"
+)
+
+// IsWorseThan returns true if this State is worse than the given State
+func (css CommitStatusState) IsWorseThan(css2 CommitStatusState) bool {
```
Comment (guillep2k) [inline modules/structs/commit_status.go]: `IsWorseOrSameAs()` maybe?

File: modules/structs/commit_status.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	switch css {
+	case CommitStatusError:
+		return true
+	case CommitStatusFailure:
+		return css2 != CommitStatusError
+	case CommitStatusWarning:
+		return css2 != CommitStatusError && css2 != CommitStatusFailure
+	case CommitStatusSuccess:
```
Comment (guillep2k) [inline modules/structs/commit_status.go]: Maybe `CommitStatusPending` should be worse than `CommitStatusSuccess`.

File: modules/structs/commit_status.go (hunk 2 above)
Comment (lunny) [reply modules/structs/commit_status.go]: Or NoBetterThan ? :)

Comment (guillep2k) [reply modules/structs/commit_status.go]: > 
> 
> Or NoBetterThan ? :)

`NoBetterThan` sounds good too. My problem was that the old name implied `<` but the function acts like `<=`.

Comment (zeripath) [reply modules/structs/commit_status.go]: Hmm... It looks like there is a genuine total order here. An alternative might be:

```go
type CommitStatus int
const (
  CommitStatusPending CommitStatus = 0
  ...
)

func (c CommitStatus) String() {
   switch c {
   case CommitStatusPending:
       return "pending"
...
   }
}

```
Then just use normal operators for the state.

File: routers/api/v1/repo/status.go [hunk 4]
```diff
@@ -220,13 +221,13 @@ func getCommitStatuses(ctx *context.APIContext, sha string) {
...
 
 type combinedCommitStatus struct {
-	State      models.CommitStatusState `json:"state"`
-	SHA        string                   `json:"sha"`
-	TotalCount int                      `json:"total_count"`
-	Statuses   []*api.Status            `json:"statuses"`
-	Repo       *api.Repository          `json:"repository"`
-	CommitURL  string                   `json:"commit_url"`
-	URL        string                   `json:"url"`
+	State      structs.CommitStatusState `json:"state"`
```
Comment (6543) [inline routers/api/v1/repo/status.go]:
```suggestion
	State      api.CommitStatusState `json:"state"`
```

File: routers/api/v1/repo/status.go [hunk 5]
```diff
@@ -11,6 +11,7 @@ import (
 	"code.gitea.io/gitea/models"
 	"code.gitea.io/gitea/modules/context"
 	"code.gitea.io/gitea/modules/repofiles"
+	"code.gitea.io/gitea/modules/structs"
```
Comment (6543) [inline routers/api/v1/repo/status.go]:
```suggestion
```

File: modules/structs/commit_status.go (hunk 2 above)
Comment (lunny) [reply modules/structs/commit_status.go]: @zeripath I will left this other PRs to keep this PR clear.

**Review by 6543 — CHANGES_REQUESTED** [review]
dont mix api&struct import ...

File: modules/structs/commit_status.go (hunk 2 above)
Comment (lunny) [reply modules/structs/commit_status.go]: @guillep2k done.

**Review by 6543 — APPROVED** [review]

**Review by guillep2k — APPROVED** [review]

**Comment by 6543** [comment]
lable: backport/done

**Pull request merged**